lib/src/contracts/contract_invoker.dart

```diff
@@ -0,0 +1,118 @@
+part of 'package:web3dart/contracts.dart';
+
+class ContractInvocation {
+  final Web3Client client;
+  final DeployedContract contract;
+  final ContractFunction function;
+  List _params;
+
+  ContractInvocation(this.client, this.contract, this.function);
+
+  ContractInvocation parameters([dynamic args]) {
+    if (args is Map) {
+      for (final p in function.parameters) {
+        if (!args.containsKey(p.name)) {
+          throw ArgumentError('Missing parameters "$p.name"');
+        }
+        _params.add(args[p.name]);
+      }
+    } else if (args is List) {
+      _params = args;
+    } else if (args != null) {
+      _params = [args];
+    }
+
+    return this;
+  }
+
+  Future<T> call<T>({
+    EthereumAddress from,
+    EtherAmount value,
+  }) {
+    return client
+        .callRaw(
+      sender: from,
+      contract: contract.address,
+      data: function.encodeCall(_params),
+      value: value,
+    )
+        .then((rawRsp) {
+      final r = function.decodeReturnValues(rawRsp);
+
+      if (T is Map) {
+        Map ret;
+
+        for (var i = 0; i < function.outputs.length; i++) {
+          final output = function.outputs[i];
+
+          ret[i] = r[i];
+
+          if (output.name != null && output.name.isNotEmpty) {
+            ret[output.name] = r[i];
+          }
+        }
+
+        return Future.value(ret as T);
+      } else if (T is List) {
+        return Future.value(function.decodeReturnValues(rawRsp) as T);
+      } else {
+        return Future.value(r.first as T);
+      }
+    });
+  }
+
+  Future<int> estimateGas({
+    @required EthereumAddress from,
+    EtherAmount value,
+  }) {
+    return client
+        .estimateGas(
+      sender: from,
+      to: contract.address,
+      value: value,
+      data: function.encodeCall(_params),
+    )
+        .then((gas) {
+      return Future.value(gas.toInt());
+    });
```

Comment:
I would use an expression function here. We also don't need to wrap this in a `Future`:
```suggestion
        .then((gas) => gas.toInt());
```